radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/core/model/CoreModelMapper.java

```diff
@@ -0,0 +1,1207 @@
+		if (subEntity == null) {
+			return new ValidatorEntity(key);
+		}
+
+		var metadata = subEntity.getMetadata();
+		if (metadata != null) {
+			throw invalidSubEntity(subEntity);
+		}
+
+		var subEntityAddress = subEntity.getAddress();
+		if (!subEntityAddress.equals(SYSTEM_ADDRESS)) {
+			throw invalidSubEntity(subEntity);
+		}
+
+		return new ValidatorSystemEntity(key);
+	}
+
+	private Entity accountAddressEntity(EntityIdentifier entityIdentifier) throws CoreApiException {
+		var address = entityIdentifier.getAddress();
+		var accountAddress = addressing.forAccounts().parseOrThrow(address, s -> invalidAddress(address));
+		var subEntity = entityIdentifier.getSubEntity();
+		if (subEntity == null) {
+			return new AccountVaultEntity(accountAddress);
+		}
+
+		switch (subEntity.getAddress()) {
+			case PREPARED_STAKES_ADDRESS -> {
+				var metadata = subEntity.getMetadata();
+				if (metadata == null || metadata.getEpochUnlock() != null || metadata.getValidatorAddress() == null) {
+					throw invalidSubEntity(subEntity);
+				}
+				var validator = addressing.forValidators().parseOrThrow(metadata.getValidatorAddress(), s -> invalidAddress(address));
+				return new PreparedStakeVaultEntity(
+					accountAddress,
+					validator
+				);
+			}
+			case PREPARED_UNSTAKES_ADDRESS -> {
+				var metadata = subEntity.getMetadata();
+				if (metadata != null) {
+					throw invalidSubEntity(subEntity);
+				}
+				return new PreparedUnstakeVaultEntity(accountAddress);
+			}
+			case EXITING_UNSTAKES_ADDRESS -> {
+				var metadata = subEntity.getMetadata();
+				if (metadata == null || metadata.getEpochUnlock() == null || metadata.getValidatorAddress() == null) {
+					throw invalidSubEntity(subEntity);
+				}
+
+				var validator = addressing.forValidators().parseOrThrow(metadata.getValidatorAddress(), s -> invalidAddress(address));
+				return new ExitingStakeVaultEntity(
+					accountAddress,
+					validator,
+					metadata.getEpochUnlock()
+				);
+			}
+			default -> throw invalidSubEntity(subEntity);
+		}
+	}
+
+	private Entity resourceAddressEntity(EntityIdentifier entityIdentifier) throws CoreApiException {
+		var address = entityIdentifier.getAddress();
+		var pair = addressing.forResources().parseOrThrow(address, s -> invalidAddress(address));
+		return new TokenEntity(pair.getFirst(), pair.getSecond());
+	}
+
+	private Entity systemAddressEntity(EntityIdentifier entityIdentifier) throws CoreApiException {
+		var subEntity = entityIdentifier.getSubEntity();
+		if (subEntity != null) {
+			throw invalidSubEntity(subEntity);
+		}
+
+		return SystemEntity.instance();
+	}
+
+	public Entity entity(EntityIdentifier entityIdentifier) throws CoreApiException {
+		var address = entityIdentifier.getAddress();
+		if (address.equals(SYSTEM_ADDRESS)) {
+			return systemAddressEntity(entityIdentifier);
+		}
+
+		var addressType = addressing.getAddressType(address).orElseThrow(() -> invalidAddress(address));
+		return switch (addressType) {
```

Comment:
Maybe we could add a TODO to use sealed classes and pattern matching in these cases?